lib/Dialect/BGV/Transforms/AddClientInterface.cpp

```diff
@@ -26,48 +26,30 @@ namespace bgv {
 #define GEN_PASS_DEF_ADDCLIENTINTERFACE
 #include "lib/Dialect/BGV/Transforms/Passes.h.inc"
 
-/// Adds the client interface for a single func. This should only be used on the
-/// "entry" func for the IR being compiled, but there may be multiple.
-LogicalResult convertFunc(func::FuncOp op, bool usePublicKey) {
-  auto module = op->getParentOfType<ModuleOp>();
-  auto argTypes = op.getArgumentTypes();
-  auto returnTypes = op.getResultTypes();
-
-  ImplicitLocOpBuilder builder =
-      ImplicitLocOpBuilder::atBlockEnd(module.getLoc(), module.getBody());
-
-  std::string encFuncName("");
-  llvm::raw_string_ostream encNameOs(encFuncName);
-  encNameOs << op.getSymName() << "__encrypt";
-
-  std::string decFuncName("");
-  llvm::raw_string_ostream decNameOs(decFuncName);
-  decNameOs << op.getSymName() << "__decrypt";
-
-  // The enryption function converts each plaintext operand to its encrypted
-  // form. We also have to add a secret key arg, and we put it at the end to
-  // maintain zippability of the non-secret-key args.
-  SmallVector<Type> encFuncArgTypes;
-  SmallVector<Type> encFuncResultTypes;
+FailureOr<lwe::RLWEParamsAttr> getRlweParmsFromFuncOp(func::FuncOp op) {
   lwe::RLWEParamsAttr rlweParams = nullptr;
+  auto argTypes = op.getArgumentTypes();
   for (auto argTy : argTypes) {
     if (auto argCtTy = dyn_cast<lwe::RLWECiphertextType>(argTy)) {
-      encFuncArgTypes.push_back(argCtTy.getUnderlyingType());
-      encFuncResultTypes.push_back(argCtTy);
       if (rlweParams && rlweParams != argCtTy.getRlweParams()) {
         op.emitError() << "Func op has multiple distinct RLWE params"
                        << " but only 1 is currently supported per func.";
         return failure();
       }
       rlweParams = argCtTy.getRlweParams();
-      continue;
     }
-
-    // For plaintext arguments, the function is a no-op
-    encFuncArgTypes.push_back(argTy);
-    encFuncResultTypes.push_back(argTy);
   }
+  return rlweParams;
```

Comment:
In case an RLWE params isn't found, does nullptr actually create a failure logicalresult? i think here an attribute can generally be a nullptr, and if it is, then there might be a segfault later.

Reply:
it looks like you catch a nullptr rlweParams on line 53 for the encryption func but not the decryption func, but you also check for the logical result failure.

Reply:
Good call. Fixed!